test/soft_rank.jl

```diff
@@ -0,0 +1,130 @@
+        @constraint(model, [i in 1:(length(x) - 1)], x[i] >= x[i + 1])
+
+        optimize!(model)
+        return value.(x)
+    end
+
+    for _ in 1:100
+        y = randn(1000)
+        x = isotonic_custom(y)
+        x2 = InferOpt.isotonic_l2(y)
+        @test all(isapprox.(x, x2, atol=1e-2))
+    end
+end
+
+@testitem "Test jacobian against finite differences" default_imports = false begin
+    include("InferOptTestUtils/src/InferOptTestUtils.jl")
+    using InferOpt, .InferOptTestUtils, FiniteDifferences, Random, Test, Zygote
+    Random.seed!(63)
+
+    for _ in 1:100
+        θ = randn(50)
+
+        sort_jac = Zygote.jacobian(x -> soft_sort_l2(x; ε=10.0), θ)[1]
+        sort_jac_fd = FiniteDifferences.jacobian(
+            central_fdm(2, 1), x -> soft_sort_l2(x; ε=10.0), θ
+        )[1]
+        @test all(isapprox.(sort_jac, sort_jac_fd, atol=1e-4))
+
+        sort_jac = Zygote.jacobian(x -> soft_sort_kl(x; ε=10.0), θ)[1]
+        sort_jac_fd = FiniteDifferences.jacobian(
+            central_fdm(2, 1), x -> soft_sort_kl(x; ε=10.0), θ
+        )[1]
+        @test all(isapprox.(sort_jac, sort_jac_fd, atol=1e-4))
+
+        rank_jac = Zygote.jacobian(x -> soft_rank_l2(x; ε=10.0), θ)[1]
+        rank_jac_fd = FiniteDifferences.jacobian(
+            central_fdm(2, 1), x -> soft_rank_l2(x; ε=10.0), θ
+        )[1]
+        @test all(isapprox.(rank_jac, rank_jac_fd, atol=1e-4))
+
+        rank_jac = Zygote.jacobian(x -> soft_rank_kl(x; ε=10.0), θ)[1]
+        rank_jac_fd = FiniteDifferences.jacobian(
+            central_fdm(2, 1), x -> soft_rank_kl(x; ε=10.0), θ
+        )[1]
+        @test all(isapprox.(rank_jac, rank_jac_fd, atol=1e-4))
+    end
+end
+
+@testitem "Learn by experience soft rank l2" default_imports = false begin
+    include("InferOptTestUtils/src/InferOptTestUtils.jl")
+    using InferOpt, .InferOptTestUtils, LinearAlgebra, Random, Test
+    Random.seed!(63)
+
+    true_encoder = encoder_factory()
+    cost(y; instance) = dot(y, -true_encoder(instance))
+    test_pipeline!(
+        PipelineLossExperience();
+        instance_dim=5,
+        true_maximizer=ranking,
+        maximizer=soft_rank,
+        loss=cost,
+        error_function=hamming_distance,
+        true_encoder=true_encoder,
+        cost=cost,
+    )
+end
+
+@testitem "Learn by experience soft rank kl" default_imports = false begin
+    include("InferOptTestUtils/src/InferOptTestUtils.jl")
+    using InferOpt, .InferOptTestUtils, LinearAlgebra, Random, Test
+    Random.seed!(63)
+
+    true_encoder = encoder_factory()
+    cost(y; instance) = dot(y, -true_encoder(instance))
+    test_pipeline!(
+        PipelineLossExperience();
+        instance_dim=5,
+        true_maximizer=ranking,
+        maximizer=soft_rank_kl,
+        loss=cost,
+        error_function=hamming_distance,
+        true_encoder=true_encoder,
+        cost=cost,
+    )
+end
+
+@testitem "Fenchel-Young loss soft rank L2" default_imports = false begin
```

Comment:
Same question as above, can we check performance gets better as we dive into the details of the CO problem to differentiate ?